contracts/cw20-ics20/src/migrations.rs

```diff
@@ -14,3 +14,43 @@ pub mod v1 {
 
     pub const CONFIG: Item<Config> = Item::new("ics20_config");
 }
+
+// v2 format is anything older than 0.13.1 when we only updated the internal balances on success ack
+pub mod v2 {
+    use crate::state::{CHANNEL_INFO, CHANNEL_STATE};
+    use crate::ContractError;
+    use cosmwasm_std::{Coin, DepsMut, Env, Order, StdResult};
+
+    pub fn update_balances(deps: DepsMut, env: &Env) -> Result<(), ContractError> {
+        let channels = CHANNEL_INFO
+            .keys(deps.storage, None, None, Order::Ascending)
+            .collect::<StdResult<Vec<_>>>()?;
+        match channels.len() {
+            0 => Ok(()),
+            1 => {
+                let channel = &channels[0];
+                let addr = &env.contract.address;
+                let states = CHANNEL_STATE
+                    .prefix(channel)
+                    .range(deps.storage, None, None, Order::Ascending)
+                    .collect::<StdResult<Vec<_>>>()?;
+                for (denom, mut state) in states.into_iter() {
+                    // this checks if we have received some coins that are "in flight" and not yet accounted in the state
+                    let Coin { amount, .. } = deps.querier.query_balance(addr, &denom)?;
+                    let diff = state.outstanding - amount;
```

Comment:
```suggestion
                    let diff = amount - state.outstanding;
```
should be cw-balance > state.outstanding ?

assuming cw-balance > state.outstanding, because state will be updated in ack (old version).

Reply:
Good catch! This is why I need tests (to panic here), or good reviewers

Reply:
still missing the handling for cw20 tokens...